crates/engine/src/operation/parse/mod.rs

```diff
@@ -1,58 +1,83 @@
 mod error;
-mod location;
+mod offsets;
 mod validation;
 
-use std::collections::HashMap;
+use std::rc::Rc;
 
-use engine_parser::{
-    types::{DocumentOperations, FragmentDefinition, OperationDefinition},
-    Positioned,
-};
+use cynic_parser::executable::OperationDefinition;
+use offsets::LineOffsets;
 use schema::Schema;
 
-pub(crate) use error::*;
-pub(crate) use location::*;
+use self::error::{ParseError, ParseResult};
+use super::Location;
+use crate::response::GraphqlError;
 
 pub(crate) struct ParsedOperation {
     pub name: Option<String>,
-    pub definition: OperationDefinition,
-    pub fragments: HashMap<engine_value::Name, Positioned<FragmentDefinition>>,
+    document: cynic_parser::ExecutableDocument,
+    line_offsets: Rc<offsets::LineOffsets>,
```

Comment:
You need a `Rc`? aren't you passing it everywhere with a ref?

Reply:
If you scroll down to the parse function you'll see the problem - I need `ParsedOperation` to own `LineOffsets` but I also need a `LineOffsets` to generate the errors if I can't build a `ParsedOperation`.

An RC seemed preferable to cloning the whole vec (even though the vec will probably be small in most cases)

Reply:
well up to you, from a quick look I think I would return `(name, document)` from the `parse_iml` and create the `ParsedOperation` + validation inside `parse`.  Seems like you would avoid the clone that way

Reply:
Way ahead of you https://github.com/grafbase/grafbase/pull/2402/commits/ad372831afa00436b533a67486f9b562f310e34f